### ISSUE TITLE
gms::_apply_state_locally_semaphore can cause unbounded memory consumption by accumulating waiters

### ISSUE THREAD

**Issue by denesb** [scylladb/scylladb#10967]
Recently we saw a customer node OOM due to 4.4M waiters on this semaphore. We should add a timeout and/or a queue limit to this semaphore to prevent unbounded memory consumption.

**Comment by xemul** [comment]
@denesb , where does this long queue comes from? Applying the state that arrives from the network should have been fixed in #5210

**Comment by xemul** [comment]
The quick fix can be to tune the way message processing waits on the `_apply_state_locally_semaphore`. Currently if the semaphore is locked the arrived message is queued and waits. The 2nd arrived one does the same. So does the Nth. However, there's no point in keeping them all with their data -- it's enough to keep only the states from the _latest_ queued message,

**Comment by denesb** [comment]
> it's enough to keep only the states from the _latest_ queued message,

For some states it is, for others it apparently [isn't](https://github.com/scylladb/scylla/pull/11030#issuecomment-1183865724).

**Comment by denesb** [comment]
> @denesb , where does this long queue comes from? Applying the state that arrives from the network should have been fixed in #5210

This is on the sender side. This issue is for the receiving side.

**Comment by xemul** [comment]
> > @denesb , where does this long queue comes from? Applying the state that arrives from the network should have been fixed in #5210
> 
> This is on the sender side. This issue is for the receiving side.

Ah, I see. So it's a node "soliciting" states from others, but failing to process them in a timely (memorily?) manner, right?

**Comment by denesb** [comment]
Yes.I think nodes can push updates to other nodes unsolicited, but I'm not exactly sure how gossip works.

**Comment by asias** [comment]
> Yes.I think nodes can push updates to other nodes unsolicited, but I'm not exactly sure how gossip works.

Gossip will only exchange the states from a peer node that it does not know about. If we queue the states and does not apply to endpoint_state_map, gossip will not consider it knows the queued states.
